algo-core/scheduler_utils.py

```diff
@@ -0,0 +1,148 @@
+
+    # Testing (define input directly):
+    # input_filename = 'support-shift-scheduler-input.json'
+
+    # Load and validate JSON input:
+    input_json = json.load(open(input_filename))
+    path_to_schema = Path(get_project_root() / 'lib/schemas/' , 'support-shift-scheduler-input.schema.json')
+    input_json_schema = json.load(open(path_to_schema))
+    try:
+        jsonschema.validate(input_json, input_json_schema)
+    except jsonschema.exceptions.ValidationError as err:
+        print("Input JSON validation error", err)
+        sys.exit(1)
+
+    input_folder = get_project_root() / 'logs' / f'{input_json["options"]["startMondayDate"]}_{input_json["options"]["modelName"]}'
+
+    return input_json
+
+def tracks_hours_to_slots(tracks):
+    for track in tracks:
+        track["start_slot"] = track["start_hour"] * 2
+        track["end_slot"] = track["end_hour"] * 2
+    return tracks
+
+def slots_to_range(week_hours, end_hour):
+    """Convert per-hour availability flags into ranges format."""
+    week_ranges = []
+
+    for day_hours in week_hours:
+        day_ranges = []
+        start = None
+
+        for i, value in enumerate(day_hours):
+            # Start of new range:
+            if start is None and value != 0:
+                start = i
+                continue
+
+            # End of range:
+            # (A range will end if either the current slot is unavailable
+            # (value 0) or if the current slot is the last one.)
+            if start is not None:
+                if value == 0:  # Unavailable
+                    day_ranges.append([start, i])
+                    start = None
+                elif i == end_hour - 1:  # Last slot
+                    day_ranges.append([start, end_hour])
+                else:
+                    continue
+
+        week_ranges.append(day_ranges)
+
+    return week_ranges
+
+
+def print_final_schedules(schedule_results, df_agents, num_days):
+    """Print final schedule, validate output JSON, and write to file."""
+    for d in range(num_days):
+        print(
+            f"\n{schedule_results[d]['start_date'].strftime('%Y-%m-%d')} "
+            "shifts:"
+        )
+
+        for (i, e) in enumerate(schedule_results[d]["shifts"]):
+            print(e)
+
+    output_json = []
+
+    for epoch in schedule_results:
+        # Substitute agent info from 'handle' to 'handle <email>'
+        shifts = []
+
+        for (name, start, end) in epoch["shifts"]:
+            shifts.append(
+                {
+                    "agent": f"{name} <{df_agents.loc[name, 'email']}>",
+                    "start": start,
+                    "end": end,
+                }
+            )
+
+        day_dict = {"start_date": epoch["start_date"].strftime("%Y-%m-%d"), "shifts": shifts}
+        output_json.append(day_dict)
+
+    output_json_schema = json.load(
+        open(Path(get_project_root() / "lib/schemas/", "support-shift-scheduler-output.schema.json")))
```

Comment:
Please see the comment above.

Reply:
I couldn't really see the signature of the constructor, I need to get used to VSCode sometimes. It makes a lot of sense now